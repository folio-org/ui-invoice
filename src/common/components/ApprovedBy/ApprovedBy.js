import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { get } from 'lodash';

import {
  KeyValue,
  NoValue,
} from '@folio/stripes/components';
import { stripesConnect } from '@folio/stripes/core';

import { BASE_RESOURCE } from '../../resources/base';
import { USERS_API } from '../../constants/api';
import { getUserName } from '../../utils';

const ApprovedBy = ({ approvedByUserId, resources }) => {
  const { failed, records } = resources.approvedByUser || {};
  const approvedByUser = failed ? null : get(records, '0');
  const value = approvedByUserId ? getUserName(approvedByUser) : <NoValue />;

  return (
    <KeyValue
      label={<FormattedMessage id="ui-invoice.invoice.approvedBy" />}
      value={value}
    />
  );
};

ApprovedBy.manifest = Object.freeze({
  approvedByUser: {
    ...BASE_RESOURCE,
    path: `${USERS_API}/!{approvedByUserId}`,
  },
});

ApprovedBy.propTypes = {
  approvedByUserId: PropTypes.string,
  resources: PropTypes.shape({
    approvedByUser: PropTypes.object,
  }).isRequired,
};

export default stripesConnect(ApprovedBy);
