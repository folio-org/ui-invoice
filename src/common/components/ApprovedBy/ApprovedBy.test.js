import React from 'react';
import { render } from '@testing-library/react';

import { ApprovedBy } from './ApprovedBy';

jest.mock('../../utils', () => ({
  getUserName: jest.fn().mockReturnValue('getUserName util'),
}));

const renderApprovedBy = (props) => (render(
  <ApprovedBy {...props} />,
));

describe('ApprovedBy', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should display no value when no approved', () => {
    const { asFragment } = renderApprovedBy({ approvedByUserId: undefined, resources: {} });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display user full name when approver defined', () => {
    const { asFragment } = renderApprovedBy({
      approvedByUserId: 'approvedByUserId',
      resources: {
        approvedByUser: { records: [{ firstName: 'Mark' }] },
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
