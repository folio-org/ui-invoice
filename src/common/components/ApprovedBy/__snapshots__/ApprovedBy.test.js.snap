// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApprovedBy should display no value when no approved 1`] = `
<DocumentFragment>
  <div
    class="kvRoot"
  >
    <div
      class="kvLabel"
    >
      ui-invoice.invoice.approvedBy
    </div>
    <div
      class="kvValue"
      data-test-kv-value="true"
    >
      <span>
        -
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ApprovedBy should display user full name when approver defined 1`] = `
<DocumentFragment>
  <div
    class="kvRoot"
  >
    <div
      class="kvLabel"
    >
      ui-invoice.invoice.approvedBy
    </div>
    <div
      class="kvValue"
      data-test-kv-value="true"
    >
      getUserName util
    </div>
  </div>
</DocumentFragment>
`;
